hal/armv7m/stm32/hal.c

```diff
@@ -0,0 +1,167 @@
+/*
+ * Phoenix-RTOS
+ *
+ * Operating system loader
+ *
+ * Hardware Abstraction Layer
+ *
+ * Copyright 2020-2021 Phoenix Systems
+ * Author: Hubert Buczynski, Marcin Baran, Gerard Swiderski, Aleksander Kaminski
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+#include <hal/hal.h>
+
+struct{
+	hal_syspage_t *hs;
+	addr_t entry;
+} hal_common;
+
+
+/* Linker symbols */
+extern void _end(void);
+extern void _plo_bss(void);
+
+/* Timer */
+extern void timer_init(void);
+extern void timer_done(void);
+
+/* Console */
+extern void console_init(void);
+
+/* Interrupts */
+extern void interrupts_init(void);
+
+
+void hal_init(void)
+{
+	hal_cpuInit();
+	_stm32_init();
+	interrupts_init();
+
+	mpu_init();
+	timer_init();
+	console_init();
+
+	hal_common.entry = (addr_t)-1;
+}
+
+
+void hal_done(void)
+{
+	timer_done();
+}
+
+
+void hal_syspageSet(hal_syspage_t *hs)
+{
+	hal_common.hs = hs;
+}
+
+
+const char *hal_cpuInfo(void)
+{
+	return "Cortex-M4 STM32L4x6";
+}
+
+
+addr_t hal_kernelGetAddress(addr_t addr)
+{
+	return addr;
+}
+
+
+void hal_kernelGetEntryPointOffset(addr_t *off, int *indirect)
+{
+	*off = sizeof(unsigned);
+	*indirect = 1;
+}
+
+
+void hal_kernelEntryPoint(addr_t addr)
+{
+	hal_common.entry = addr;
+}
+
+
+int hal_memoryAddMap(addr_t start, addr_t end, u32 attr, u32 mapId)
+{
+	return mpu_regionAlloc(start, end, attr, mapId, 1);
+}
+
+
+static void hal_getMinOverlappedRange(addr_t start, addr_t end, mapent_t *entry, mapent_t *minEntry)
+{
+	if ((start < entry->end) && (end > entry->start)) {
+		if (start > entry->start)
+			entry->start = start;
+
+		if (end < entry->end)
+			entry->end = end;
+
+		if (entry->start < minEntry->start) {
+			minEntry->start = entry->start;
+			minEntry->end = entry->end;
+			minEntry->type = entry->type;
+		}
+	}
+}
+
+
+int hal_memoryGetNextEntry(addr_t start, addr_t end, mapent_t *entry)
+{
+	mapent_t tempEntry = { 0 };
+	mapent_t minEntry = { .start = (addr_t)-1, .end = 0 };
```

Comment:
What if the compiler decides to use the `memset()` function to reset the value ( as with -Os option ).

Reply:
Fixed other examples too, thank you